rustc_resolve: fix instability in lib.rmeta contents

durin42 opened the pull request:
rust-lang/rust@23032f31c91f2 accidentally introduced some nondeterminism in the ordering of lib.rmeta files, which we caught in our bazel-based builds only recently due to being further behind than normal. In my testing, this fixes the issue.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

durin42 committed:
rustc_resolve: fix instability in lib.rmeta contents

rust-lang/rust@23032f31c91f2 accidentally introduced some nondeterminism
in the ordering of lib.rmeta files, which we caught in our bazel-based
builds only recently due to being further behind than normal. In my
testing, this fixes the issue.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13932995200/job/38994508261) [(plain)](https://github.com/rust-lang/rust/commit/8b6d60d04151f2b60ef9ef1fbc0afeb8d80244d3/checks/38994508261/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.1s done
#19 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

workingjubilee commented on compiler/rustc_resolve/src/rustdoc.rs:
```diff
@@ -454,7 +455,7 @@ fn parse_links<'md>(doc: &'md str) -> Vec<Box<str>> {
     }
 
-    for (label, refdef) in event_iter.reference_definitions().iter() {
+    for (label, refdef) in event_iter.reference_definitions().iter().sorted_by_key(|x| x.0) {
```
Please just change the FxHashSet to a BTreeSet.

workingjubilee replied:
Oh, wait, this is a HashMap in another crate... nevermind.

fmease commented:
Sorry for the delay, thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit 795a6669f83274ce3ada1046e6c9284b8e900a98 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 795a6669f83274ce3ada1046e6c9284b8e900a98 -->
<!-- homu: {"type":"Approved","sha":"795a6669f83274ce3ada1046e6c9284b8e900a98","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

fmease commented:
Wait, could you also update the `reference_definitions` in `src/librustdoc/html/markdown.rs`?

@bors r-

fmease commented on compiler/rustc_resolve/src/rustdoc.rs:
```diff
@@ -454,7 +455,7 @@ fn parse_links<'md>(doc: &'md str) -> Vec<Box<str>> {
 
-    for (label, refdef) in event_iter.reference_definitions().iter() {
+    for (label, refdef) in event_iter.reference_definitions().iter().sorted_by_key(|x| x.0) {
         if !refids.contains(label) {
```
Btw, if you want to, you could additionally harden this by swapping out the `FxHashSet` (the type of `refids`) with `UnordSet` which is just a wrapper around the former that forbids any sort of iteration

durin42 replied:
Sure

durin42 committed:
librustdoc: also stabilize iteration order here

durin42 commented:
> Wait, could you also update the callsite of `reference_definitions` in `src/librustdoc/html/markdown.rs`?
> 
> @bors r-

Done

durin42 committed:
rustc_resolve: prevent iteration of refids for completeness

This came up in review, and it should help some future author
not introduce non-deterministic output here.

fmease commented:
@bors r+

bors commented:
:pushpin: Commit 1437dec79942f1367b4cc7756e2f080261c34f3a has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 1437dec79942f1367b4cc7756e2f080261c34f3a -->
<!-- homu: {"type":"Approved","sha":"1437dec79942f1367b4cc7756e2f080261c34f3a","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
